pkg/caveats/types/typeset.go

```diff
@@ -0,0 +1,71 @@
+package types
+
+import (
+	"fmt"
+
+	"github.com/authzed/cel-go/cel"
+)
+
+// TypeSet defines a set of types that can be used in caveats. It is used to register custom types
+// and methods that can be used in caveats. The types are registered by calling the RegisterType
+// function. The types are then used to build the CEL environment for the caveat.
+type TypeSet struct {
+	// definitions holds the set of all types defined and exported by this package, by name.
+	definitions map[string]typeDefinition
+
+	// customTypes holds the set of custom types defined and exported by this package. This is exported
+	// so that the CEL environment construction can apply the necessary env options for the custom
+	// types.
+	customTypes map[string][]cel.EnvOption
+
+	// customMethodsOnTypes holds a set of new methods applied over defined types. This is exported
+	// so that the CEL environment construction can apply the necessary env options to support these methods
+	customMethodsOnTypes []cel.EnvOption
+
+	// isFrozen indicates whether the TypeSet is frozen. A frozen TypeSet cannot be modified.
+	isFrozen bool
+}
+
+// Freeze marks the TypeSet as frozen. A frozen TypeSet cannot be modified.
+func (ts *TypeSet) Freeze() {
+	ts.isFrozen = true
+}
```

Comment:
Is this a convention? Where does this actually prevent the TypeSet from being modified?

Reply:
Check below: it can't be used if not frozen

Reply:
Right, it can't be used, but nothing prevents me from touching a frozen typeset (unless I'm missing something)